app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-core.js

```diff
@@ -273,6 +273,27 @@ window.GOVUK.analyticsGa4 = window.GOVUK.analyticsGa4 || {};
               index_link: parseInt(index[2])
             }
         }
+      },
+
+      addAttributesToElements: function (targetElementsSelector, dataAttributes) {
+        var targetElements = document.querySelectorAll(targetElementsSelector)
+
+        for (var i = 0; i < targetElements.length; i++) {
+          var element = targetElements[i]
+
+          for (var j = 0; j < dataAttributes.length; j++) {
+            var key = dataAttributes[j].key
+            var value = dataAttributes[j].value
```

Comment:
If I'm understanding this code correctly, it looks like we're depending on `key` and `value` existing on `dataAttributes[j]` - I've been (possibly over) thinking this and I'm wondering if setting `var key` and `var value` to something like `dataAttributes[j][Object.keys(dataAttributes[j])[0]]` and `dataAttributes[j][Object.keys(dataAttributes[j])[1]]` respectively would make things more robust as it wouldn't matter what the properties on `dataAttributes[j]` are called.

Having said that, I don't think it's a very readable solution and it could be hard to understand. Alternatively, do you think another option could be wrapping this code in another guard condition - something like `if(dataAttributes[j].hasOwnProperty('key') && dataAttributes[j].hasOwnProperty('value')) { ... }`?

Reply:
Thanks - I think the second alternative you suggested is probably worth adding. 

I am wondering about our wider approach to "checking data passed to a function is correct" though, because the function would also crash if someone passed a number or object to the `selector` value - and if we were checking for that as well to prevent crashes, functions would start getting big.

My personal view is that when the data passed into a function is fully controlled by developers, it should be OK not to do any strict checking. This is because I think a developer should be responsible enough to ensure that what they're passing into a function is correct. However, when the data being passed to a function is unreliable (e.g. if the data passed into the function is coming from the DOM or an API), then I think there should be stricter checks. What do you think?

@andysellick Do you have any views on strict input checking on functions? I couldn't find one in our JS coding style guide.

Reply:
I don't think we've ever written down anything to this effect, but your approach of the level of checking being dependent on where the data is coming from sounds sensible.